### PR TITLE
support OIDC claims request

### LINKED ISSUE #1956: Request OIDC groups claim if groups scope is not supported
**Is your feature request related to a problem? Please describe.**
Not all OIDC providers support a groups scope (like OneLogin and Microsoft) and do not return the groups claim as part of the ID token. Still, these OIDC providers might support a custom 'groups' claim if explicitly instructed to do so.

**Describe the solution you'd like**
Based on the OIDC config it is possible to check whether or not the OIDC provider supports the groups scope. If the groups scope is not supported, add an explicit 'groups' claim request for the ID token.

### PR DESCRIPTION
**Pull request by sboschman**
Adds a ID token 'groups' claim request if the OIDC provider does not support the groups scope.

fixes #1956

### CODE REVIEW

**Comment by jessesuen** [comment]
@sboschman this was supposed to be handled by the following setting, to control if it is requested:
https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/argocd-cm.yaml#L45

And this setting to control if it is enforced:
https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/argocd-rbac-cm.yaml#L26

So do we really need this?

**Comment by sboschman** [comment]
@jessesuen, the settings you linked are only to control the requested scopes as far as I can tell, not to make claim requests. Not all OIDC providers support this 'groups' scope which returns a groups claim as part of the ID token. In fact the groups scope is not part of the oidc specification, so oidc providers are not required to support it.

That doesn't mean though that an OIDC provider which does not support the groups scope can not return custom defined claims, e.g. a groups claim. The oidc specification states that such claims should be returned on the userinfo endpoint by default (and are not returned as part of the ID token).

Flow: get ID Token + AccessToken -> use AccessToken to query UserInfo endpoint -> get custom claims from UserInfo

The oidc specification allows claim requests to be part of the authcode flow, requesting to return certain claims as part of the ID token. So if the oidc provider does not support the groups scope, it might still be able to return the groups claim if instructed to do so with a claim request.

E.g. to get groups working with my OIDC provider, I have to request the scopes: openid profile groups
(because it does not know 'groups' as a scope it will treat it as a custom claim). To have it return the groups claim as part of the ID token I have to supply a claim request (https://github.com/argoproj/argo-cd/pull/1957/files#diff-d5680ae5ff2ca428c44aff82001d6345R267).

I hope this explains that this change is supplemental to the settings you linked, and not a duplicate of it.

**Comment by jessesuen** [comment]
Thanks for the explanation. I've done some more reading, and I think I am starting to understand. Are you basically saying that today, Argo CD supports requesting claims only using the scopes parameter:
https://connect2id.com/products/server/docs/guides/requesting-openid-claims#scope-parameter

But with this PR, you would additionally be supporting using the claims parameter?
https://connect2id.com/products/server/docs/guides/requesting-openid-claims#claims-parameter

If so, I think this would be a valid enhancement. However, I think may be noticing room for improvement:

1. If we want to support retrieving claims using the claims parameter, I think we should generalize this and add something to the config (e.g. `requestedClaims`) to let user's customize this, just like we do for `requestedScopes`. This way, it's not hard-wiring `groups` into the claims request, and giving the user control.

2. I see that the change only affected the GrantTypeAuthorizationCode case but not GrantTypeImplicit. The latter is used for CLI logins, and I think the solution may be incomplete there. In other words, I believe we would still want to adjust the url to include requested claims for the GrantTypeImplicit code path.

**Comment by sboschman** [comment]
Yes, that is correct @jessesuen about the existing scope parameter support and additionally trying to support the claims parameter.

In regards to your first point:
I am a bit reluctant to add more complex configuration, all though I agree the solution should maybe be a bit more flexible. Perhaps you can have a look at the new code I submitted. Hopefully this is an acceptable alternative to adding more configuration options.
In short it takes the requested scopes from the config, removes all the supported scopes from the OIDC provider, and adds a ID token claim request for the remaining scopes/claims.

In regards to your second point:
Will have to look into this. CircleCI did not approve of my earlier attempt to add a claim request to the implicit flow.

**Comment by jessesuen** [comment]
Although it may be slightly more complex to configure, I think it's the right thing to do. Looking at the openid spec:
https://openid.net/specs/openid-connect-core-1_0.html#ClaimsParameter

It seems we would be shutting the door on the ability to use some of the other features of the spec, that people may want to use in ways that we cannot anticipate. For example, the current PR hard-wires the claim as `essential: true`, but does not provide any way to adjust requesting of the `values` array.

I think providing the maximum amount of flexibility at a one-time slight increase in cost in configuration is worth the trouble.

Also, as I was thinking about documenting the behavior of the current PR, it seems it is hard to explain. Basically, we have a single configuration field, `requestedScopes`, which doubles both as the scopes parameter and claims parameter, with the caveat that if the scope is not supported, it would be used as a claims parameter, marking it as essential. This is something that is pretty non-standard. Compare that to the simplicity of explaining that `requestedScopes` configures the scopes parameter in an OIDC request, and `requestedClaims` configures the claims parameters in an OIDC request.

**Comment by sboschman** [comment]
We can add extra configuration of course. Can we simplify your earlier configuration though?
```
    requestedClaims: |
        {
            "id_token" : {
                "email"          : { "essential" : true },
                "email_verified" : { "essential" : true },
            },
            "userinfo" {
                "email"          : { "essential" : true },
                "email_verified" : { "essential" : true },
                "name"           : { "essential" : true },
            }
        }
```
This config covers the complete OIDC spec. Argo CD never queries the userinfo endpoint, so the "userinfo" part of the claim request is pointless. Specifying the claims that should be returned on the "id_token" part is enough, imo.

```
   requestedClaims: |
      {
                "email"             : { "essential" : true },
                "email_verified"    : { "essential" : true },
                "groups"            : { "essential": true, "value": "org:myorg" },
                "acr"               : {"essential": true,
                                       "values": ["urn:mace:incommon:iap:silver",
                                       "urn:mace:incommon:iap:bronze"]}
      }
```
In the documentation we can state that the configured claims are returned on the ID token. As Argo CD will never query the userinfo endpoint, so all required claims should be available on the ID token.

Any thoughs on this @jessesuen  ?

**Comment by jessesuen** [comment]
I think we can simplify it that way, but then let's rename the key from `requestedClaims` to `requestedIDTokenClaims`.

Another thought I had was, instead of a literal json string, we could make it structured data. So it would then be possible to do this:

```yaml
   requestedIDTokenClaims:
     email:
       essential: true
     email_verified:
       essential: true
     groups:
       essential: true
       value: org:myorg
     acr:
       essential: true
       values:
       - urn:mace:incommon:iap:silver
       - urn:mace:incommon:iap:bronze
```

Or in the simple case (below is valid YAML):
```yaml
   requestedIDTokenClaims: {"groups": {"essential": true}}
```

**Comment by jessesuen** [comment]
Basically this is what I suggest:

```golang
type IDTokenClaim struct {
    Essential bool
    Value string
    Values []string
}

type OIDCConfig struct {
...
	RequestedIDTokenClaims map[string]IDTokenClaim `json:"requestedIDTokenClaims,omitempty"`
...
}

File: util/oidc/oidc.go [hunk 1]
```diff
@@ -175,13 +186,14 @@ func (a *ClientApp) HandleLogin(w http.ResponseWriter, r *http.Request) {
...
+
 	grantType := InferGrantType(oauth2Config, oidcConf)
 	var url string
 	switch grantType {
 	case GrantTypeAuthorizationCode:
-		url = oauth2Config.AuthCodeURL(stateNonce)
+		url = oauth2Config.AuthCodeURL(stateNonce, opts...)
 	case GrantTypeImplicit:
-		url = ImplicitFlowURL(oauth2Config, stateNonce)
+		url = ImplicitFlowURL(oauth2Config, stateNonce, opts...)
```
Comment (jessesuen) [inline util/oidc/oidc.go]: Even though we pass the `AuthCodeOption` option is passed to `ImplicitFlowURL`, `ImplicitFlowURL` logic is not modifying the URL like AuthCodeURL is doing. See:

```golang
	for _, opt := range opts {
		switch opt {
		case oauth2.AccessTypeOnline:
			v.Set("access_type", "online")
		case oauth2.AccessTypeOffline:
			v.Set("access_type", "offline")
		case oauth2.ApprovalForce:
			v.Set("approval_prompt", "force")
		}
		// I think we need else case here to set the URL
	}
```

Can you fix `ImplicitFlowURL` to also modify the URL so that this will work for CLI logins? To test this, from Argo CLI you can rebuild argocd CLI, run `argocd login --sso` and verify the generated URL has the claims URL parameter.

**Review by jessesuen — CHANGES_REQUESTED** [review]

File: util/oidc/oidc.go (hunk 1 above)
Comment (sboschman) [reply util/oidc/oidc.go]: Yes I am trying to solve it. It gets a bit complicated though with the protobuf settings vs the normal settings, reusing code in the util package, unallowed import cycles and my lack of go skills.

**Comment by sboschman** [comment]
When you have time, please have a look at it @jessesuen .

I changed the protobuf settings so the requested claims setting is also available in login.go/relogin.go to pass to the oidc call. This oauth2.AuthCodeOption is pretty annoying with all the private stuff, so I ended up changing the ImplicitFlowURL function, see https://github.com/argoproj/argo-cd/pull/1957/files#diff-d5680ae5ff2ca428c44aff82001d6345R310. Basically using the oauth2.AuthCodeURL function and overwriting the response type to "id_token", and adding the nonce.

**Comment by alexec** [comment]
@jessesuen any update on this?

**Review by jessesuen — APPROVED** [review]
This looks good to go. Just resolve conflicts and we'll merge it in. Thanks!

**Comment by sboschman** [comment]
@jessesuen conflicts are resolved after rebase on master

**Comment by alexmt** [comment]
Thank you @sboschman !
